libs/dyn/pattern_trie.go

```diff
@@ -0,0 +1,182 @@
+package dyn
+
+import "fmt"
+
+// PatternTrie is a trie data structure for storing and querying patterns.
+// It supports both exact matches and wildcard matches. You can insert [Pattern]s
+// into the trie and then query it to see if a given [Path] matches any of the
+// patterns.
+type PatternTrie struct {
+	root *trieNode
+}
+
+// trieNode represents a node in the pattern trie.
+// Note that it can only be one of anyKey, anyIndex, collection of pathKeys, or collection of pathIndexes.
```

Comment:
I don't think this statement is enforced well enough in the current code, which is prone to state inconsistencies, which are difficult to debug, if multiple fields end up being set?

Reply:
Good point. That comment is misleading and I've removed it. Also added a bigger blurb explaining it all. It is valid for both anyKey and pathKeys to be set at the same time and we have a test that actually does so.

From the perspective of the trie it's also fine if both anyKey and anyIndex are set at the same time, but since such a scenario does not arise in practice I've omitted to support that in `SearchPath`.

While validation would be nice to have, it seem unnecessary to me given the unit tests already cover correctness. I'm happy to add validation that both key (one of anyKey and pathKeys) and index (one of anyIndex and pathIndex) are not set at the same time if you think we should add it regardless.